src/Perlang.Interpreter/Typing/TypeResolver.cs

```diff
@@ -388,6 +389,75 @@ public override VoidObject VisitCallExpr(Expr.Call expr)
             return VoidObject.Void;
         }
 
+        public override VoidObject VisitIndexExpr(Expr.Index expr)
+        {
+            base.VisitIndexExpr(expr);
+
+            Type? type = expr.Indexee.TypeReference.ClrType;
+            Type? argumentType = expr.Argument.TypeReference.ClrType;
+
+            if (argumentType == null)
+            {
+                // TODO: Be ITokenAware and use expr.Argument.Token in this error, for better accuracy
+                typeValidationErrorCallback(new TypeValidationError(
+                    expr.Token,
+                    "Internal compiler error: Type of index argument expected to be resolved at this stage")
+                );
+
+                return VoidObject.Void;
+            }
+
+            switch (type)
+            {
+                case { } when type == typeof(string):
+                    if (!argumentType.IsAssignableTo(typeof(int)))
+                    {
+                        typeValidationErrorCallback(new TypeValidationError(
+                            expr.ClosingBracket,
+                            $"'string' cannot be indexed by '{argumentType.ToTypeKeyword()}'")
+                        );
+                    }
+
+                    expr.TypeReference.ClrType = typeof(char);
+                    break;
+
+                // TODO: arrays
+                // TODO: support lists
```

Comment:
This will be left unsupported for now. We don't have a way to define user-defined arrays or lists in the language, and I'm not aware of any `stdlib` method that currently needs this. :thinking: We can easily add it when we see the need for it, and I don't want to add support for it without zealously adding integration tests for it at the same time... :slightly_smiling_face: 

Will keep #270 open precisely because of this.